### Data processing  
**mRNA analysis**    
* Microarray. For Affymetrix expression arrays (HG-U133_Plus_2, HG-U133A, HG_U95Av2 and HuGene-1_0-st) the raw .CEL files (when available) were downloaded from the respective sources and the analyses were completed in R using the [Bioconductor](http://bioconductor.org) suite. The [‘affy’](http://www.bioconductor.org/packages/release/bioc/manuals/affy/man/affy.pdf) package was used for robust multi-array average normalization followed by quantile normalization. For genes with several probe sets, the median of all probes had been chosen. In case the raw data were not available and for other expression array platforms (Illumina beadchip, Agilent-4502A), the available pre-processed expression matrixs were used.            
* RNA-seq. The normalized count reads from the pre-processed data (sequence allignment and transcript abundance estimation) were log2 transformed after adding a 0.5 pseudocount (to avoid infinite value upon log transformation).    

**Copy number**     
* Copy number alterations for the TCGA datasets (obtained through the [GISTIC 2.0](http://www.broadinstitute.org/cgi-bin/cancer/publications/pub_paper.cgi?mode=view&paper_id=216&p=t) algorithm) were downloaded from the [Broad GDAC Firehose](http://gdac.broadinstitute.org).    

**Survival**         
* Kaplan–Meier survival analysis are performed using the [‘survival’](http://cran.r-project.org/web/packages/survival/survival.pdf) package in R. Hazard ratios were determined utilizing the `coxph` function from the ‘survival’ package. To generate HR plot we partially use some code previously describend in [*Cutoff finder*](http://molpath.charite.de/cutoff/index.jsp).    

**Mutations**  
* Mutation data (identified with the MutSig2CV algorithm) are dinamically downloaded from the cBio Portal using the [‘cgdsr’](https://cran.r-project.org/web/packages/cgdsr) package.      

***
### Classification     
**CIMP status**    
CIMP status has been determined  by **Support Vector Machine (SVM)**, using TCGA GBM as training dataset.    

**Subtypes**   
*- GBM*   
For the GBM subtype classification we use the 3 TCGA expression subtypes (Classical, Mesenchymal,Proneural) as defined in [Wang Q. et al. 2016](http://biorxiv.org/content/early/2016/08/13/052076).   
   
*- LGG*  
For the LGG subtype classification we use the 3 TCGA molecular subtypes (IDH mutant with 1p/19q codeletion, IDH mutant without 1p/19q codeletion, IDH wild-type) as defined in [NEJM, 2015](http://www.ncbi.nlm.nih.gov/pubmed/26061751). **Important note:** the GlioVis LGG classification identifies "molecular subtypes" using expression data. Such approach has to be considered very preliminary untill further validation.        
   
Classification algorithms:      
* The Subtype calls generated by **Support Vector Machine (SVM)**, are done performing a linear Kernel with 10-fold cross validation, using the `ksvm` function of the [‘kernlab’](http://cran.r-project.org/web/packages/kernlab/kernlab.pdf) package. As training dataset for the GBM we use the TCGA GBM samples described in [Wang Q. et al. 2016](http://biorxiv.org/content/early/2016/05/08/052076). As training dataset for the LGG we use the TCGA LGG samples described in [NEJM](http://www.ncbi.nlm.nih.gov/pubmed/26061751).         
* The **K-nearest neighbor (K-NN)** Subtype calls, are obtained using the `knn3Train` function of the [‘caret’](http://cran.r-project.org/web/packages/caret/caret.pdf) package, using the subtypes as in the SVM calls.  
* The Subtype call generated  by **single sample Gene Set Enrichment Analysis (ssGSEA)**, is done using the `runSsGSEAwithPermutation` function of the [‘ssgsea.GBM.classification’](http://biorxiv.org/content/early/2016/05/08/052076) package. Not available yet for LGG samples.   
 
***
### App info

Version 0.20 (18/03/2016)

By: [Massimo Squatrito](http://www.cnio.es/ing/grupos/plantillas/presentacion.asp?grupo=50008938)

Built using [Shiny](http://www.rstudio.com/shiny) by [RStudio](http://www.rstudio.com)

Code available on [GitHub](https://github.com/msquatrito/shiny_GlioVis)

### License
GlioVis is free software: you can redistribute it and/or modify it under the terms of the [GNU Affero General Public License](http://opensource.org/licenses/GPL-3.0), version 3, as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details. 

&copy; Massimo Squatrito (2020) <a rel="license" href="http://www.creativecommons.org/licenses/by-nc/4.0/"><img alt="Creative Commons License" style="border-width:0" src="cc.png" /></a><br />This work is licensed under a <a rel="license" href="http://www.creativecommons.org/licenses/by-nc/4.0/">Creative Commons Attribution-NonCommercial 4.0 International License</a>. 

***
### Acknowledgments  
We are very grateful to all the laboratories that have generated and shared the data used in this web application. We would like to thank Angel Carro and David Pisano, from the Spanish National Research Center (CNIO), for the support in deploying the application.

  
### Funding

<a href ="http://www.seveballesteros.com"><img src="seve.jpg" height = "135"  width = "405"></a>

